src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -102,6 +89,113 @@ pub struct ConfigSpace {
+    // Buffers that we have used and they are ready to be given back to the guest.
+    pub deferred_descriptor: Option<ParsedDescriptorChain>,
+}
+
+impl RxBuffers {
+    /// Create a new [`RxBuffers`] object for storing guest memory for performing RX
+    fn new() -> Result<Self, IoVecError> {
+        Ok(Self {
+            min_buffer_size: 0,
+            iovec: IoVecBufferMut::new()?,
+            parsed_descriptors: VecDeque::with_capacity(FIRECRACKER_MAX_QUEUE_SIZE.into()),
+            deferred_descriptor: None,
+        })
+    }
+
+    /// Add a new `DescriptorChain` that we received from the RX queue in the buffer.
+    ///
+    /// SAFETY: The `DescriptorChain` cannot be referencing the same memory location as any other
+    /// `DescriptorChain`. (See also related comment in
+    /// [`IoVecBufferMut::append_descriptor_chain`]).
+    unsafe fn add_buffer(
+        &mut self,
+        mem: &GuestMemoryMmap,
+        head: DescriptorChain,
+    ) -> Result<(), AddRxBufferError> {
+        let parsed_dc = self.iovec.append_descriptor_chain(mem, head)?;
+        if parsed_dc.length < self.min_buffer_size {
+            self.iovec.drop_descriptor_chain(&parsed_dc);
+            return Err(AddRxBufferError::BufferTooSmall);
+        }
+        self.parsed_descriptors.push_back(parsed_dc);
+        Ok(())
+    }
+
+    /// Returns the number of available `iovec` objects.
+    #[inline(always)]
+    fn len(&self) -> usize {
+        self.iovec.len()
+    }
+
+    /// Mark the first `size` bytes of available memory as used.
+    ///
+    /// The `size` should be the number of bytes previously written in the [`RxBuffers`]. This
+    /// means that `size` cannot be more than the currently available bytes in the underlying
+    /// `IoVecBufferMut`.
+    fn mark_used(&mut self, size: u32, metrics: &NetDeviceMetrics) {
+        // Since we were able to write a frame in guest memory, we should have at least one
+        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
+        // fundamentally broken.
+        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
+            "net: internal bug. Mismatch between written frame size and available descriptors",
+        );
+
+        // We are trying to mark as used more bytes than a single descriptor. However without
+        // VIRTIO_NET_F_MRG_RXBUF negotiated (we don't support it currently) the guest can't handle
+        // frames spanning multiple DescriptorChains. There is nothing we can do now, just pass the
+        // descriptor back to the guest without claiming any written bytes
+        if size > parsed_dc.length {
+            parsed_dc.length = 0;
+            metrics.rx_fails.inc();
+        } else {
+            parsed_dc.length = size;
+            metrics.rx_bytes_count.add(size as u64);
+            metrics.rx_packets_count.inc();
+        }
+        self.iovec.drop_descriptor_chain(&parsed_dc);
```

Comment:
Can you also add an update of the packet header with correct number of buffers like it is done here: https://github.com/firecracker-microvm/firecracker/pull/4813/files#diff-205f89a698c2fafa14dcbd3c76dfee31dd597294c8334bc6b82d8c92814a497fR260
This is part of the spec and without MRG_RXBUF the `num_buffers` should always be 1.

Reply:
added this.